### ISSUE TITLE
How should I implement the baseline  with the accuracy 77.9% in VGG19?

### ISSUE THREAD

**Issue by lanfeng4659** [jeong-tae/RACNN-pytorch#4]
hello, I'm interesting in RACNN too. And I'm writing the baseline with pytorch. however, I get the highest accuracy of 73.18%. so, have you implemented the performance reported in his paper? waiting for your reply. thank you!

**Comment by jeong-tae** [comment]
hello. you can use the model that implemeted in pytorch maybe if you implemented with the RACNN verseion, you will be fail to get highest score.

i am not sure that how did they get the results exactly. i wrote the result following the paper

**Comment by yuleichin** [comment]
@jeong-tae  Hi, I have some questions about the code. In the crop layer, there is a slicing operation x[tx-tl:tx+tl, ty-tl:ty+tl] to crop the target area. Then, the cropped image is resampled using nn.upsample function.
However, if we directly crop the target region using slicing operation on tensors, why we still need to multiply a mask (boxcar function) with the original image?
Is there exists a possibility that we directly bilinear-upsample the image based on the results of multiplying the mask with original image? (-e.g., use the equation 7 in the paper)

**Comment by jeong-tae** [comment]
@yuleichin This is not intended code by myself. It's on the paper, eq(4), (5), (6).
They says there are two reason: Approximate the cropping operation and Analytical representation which is necessary when optimizing box parameters in backward-propagation.

...But i also think that is it necessary to operate? when i remove the mask with original, got the almost same result. Still don't know what happen in the code.
If you interested in this, you can fork the code and try without masking.

**Comment by yuleichin** [comment]
@jeong-tae I would like to ask one question about the backward() function in your code:

**return None, ret**
As suggested in the website [https://pytorch.org/docs/stable/notes/extending.html](url), why not **return ret, None**?
Is it the reason why your loss of apn doesn't decrease?

**Comment by jeong-tae** [comment]
I can not find any suggestion about parameter order to return. Can you point out exact mention about it?

i return the gradient w.r.t its corresponding inputs. images, and apn loc.

is there a mistake? hm...

**Comment by yuleichin** [comment]
Sorry for that, your input arguments' order is different from the RACNN code in DeepCar project. I think it's OK....But I find that in my project, when the epoch increases, the GPU memory continues to grow. Maybe I should delete some reference in the backward propagation?

**Comment by jeong-tae** [comment]
i don't know about detail of your code so that it's hard to help you.
Please carefully review your code and find the where the leak occur. Useless code should be deleted!

Also look around the data loader or iteration that use CUDA, variables.

**Comment by yuleichin** [comment]
@jeong-tae Thank you very much~ I rewrite the back propagation function. Since my torch.version is 0.3, so the grad_output is a Variable in backward(). Thus, all operations are performed on Variables instead of tensors. Besides, after each training epoch, I use torch.cuda.empty_cache() to release some GPU memory. My code is as the following:
`class AttentionCropFunction(Function):
    @staticmethod
    def forward(ctx,inputs,apn_out):
        h = lambda x: 1 / (1 + torch.exp(-10 * x))
        in_size = inputs.size(-1)
        in_channel = inputs.size(1)
        assert (in_channel == 3 or in_channel == 1)
        unit = torch.stack([torch.arange(0, in_size)] *in_size) #256
        x = torch.stack([unit.t()] * in_channel)
        y = torch.stack([unit] * in_channel)
        if isinstance(inputs, torch.cuda.FloatTensor) or\
        isinstance(inputs, torch.cuda.FloatTensor):
            x, y = x.cuda(), y.cuda()
        #print ('type inputs, apn_out ,size ',type(inputs),type(apn_out),inputs.size(),apn_out.size())
        #<class 'torch.cuda.FloatTensor'>, <class 'torch.cuda.FloatTensor'>)
        # (5L, 3L, 256L, 256L) (5L, 3L),
        #Now the parameters are all tensor, not Variables
        ret = []
        batchlen = inputs.size(0)
        
        for i in range(batchlen):
            tx, ty, tl = apn_out[i][0], apn_out[i][1], apn_out[i][2]
            #assert tl >= 0
            tx = tx if tx > (in_size/3) else in_size/3
            tx = tx if (in_size/3*2) < tx else (in_size/3*2)
            ty = ty if ty > (in_size/3) else in_size/3
            ty = ty if (in_size/3*2) < ty else (in_size/3*2)
            tl = tl if tl > (in_size/3) else in_size/3

            w_off = int(tx-tl) if (tx-tl) > 0 else 0
            h_off = int(ty-tl) if (ty-tl) > 0 else 0
            w_end = int(tx+tl) if (tx+tl) < in_size else in_size
            h_end = int(ty+tl) if (ty+tl) < in_size else in_size

            mk = (h(x-w_off) - h(x-w_end)) * (h(y-h_off) - h(y-h_end))
            assert(mk.size(0)==in_channel)
            xatt = inputs[i] * mk
            xatt_cropped = xatt[:, h_off : h_end, w_off : w_end].contiguous()
            before_upsample = Variable(xatt_cropped.unsqueeze(0))
            xamp = F.upsample(before_upsample, size=(128,128), mode='bilinear') # upsample to 128
            assert (xamp.size(-1)==128)
            ret.append(xamp.data.squeeze(0))
            del before_upsample, xamp
        ret_tensor = torch.stack(ret)
        #print ('ret tensor size, type', ret_tensor.size(), type(ret_tensor))
        #('ret tensor size, type', (5L, 3L, 128L, 128L), <class 'torch.cuda.FloatTensor'>)
        assert (ret_tensor.size(0)==batchlen and ret_tensor.size(1)==in_channel)
        del x, y, unit
        ctx.save_for_backward(inputs, apn_out)
        return ret_tensor
        
    @staticmethod
    def backward(ctx, grad_output):
        inputs, apn_out = ctx.saved_variables
        grad_input = grad_apn = None
        
        grad_apn = Variable(torch.Tensor(grad_output.size(0), 3).zero_())
        if isinstance(grad_output.data, torch.cuda.LongTensor) or\
        isinstance(grad_output.data, torch.cuda.FloatTensor):
            grad_apn = grad_apn.cuda()
        
        norm = -(grad_output * grad_output).sum(dim=1)
        
        ##print grad_output
        #show_image(inputs.cpu().data[0])
        #show_image(ret_tensor.cpu().data[0])
        #plt.imshow(norm[0].cpu().numpy(), cmap='gray')
        in_size=128
        x = Variable(torch.stack([torch.arange(0, in_size)] * in_size).t())
        y = x.t()
        #print (x, y) 'torch.autograd.variable.Variable
        #mx = (x >= 168).float() - (x < 56).float()
        #my = (y >= 168).float() - (y < 56).float()
        #ml = (((x<56)+(x>=168)+(y<56)+(y>=168)) > 0).float()*2 - 1
        long_size = int(in_size/3*2)
        short_size = int(in_size/3)
        mx = (x >= long_size).float() - (x < short_size).float()
        my = (y >= long_size).float() - (y < short_size).float()
        ml = (((x<short_size)+(x>=long_size)+(y<short_size)+(y>=long_size)) > 0).float()*2 - 1        
        mx_batch = torch.stack([mx.float()] * grad_output.size(0))
        my_batch = torch.stack([my.float()] * grad_output.size(0))
        ml_batch = torch.stack([ml.float()] * grad_output.size(0))
        #type(mx_batch),type(my_batch),type(ml_batch)'torch.autograd.variable.Variable
        if isinstance(grad_output.data, torch.cuda.FloatTensor) or\
        isinstance(grad_output.data, torch.cuda.LongTensor):
            mx_batch = mx_batch.cuda()
            my_batch = my_batch.cuda()
            ml_batch = ml_batch.cuda()

        grad_apn[:, 0] = (norm * mx_batch).sum(dim=1).sum(dim=1)
        grad_apn[:, 1] = (norm * my_batch).sum(dim=1).sum(dim=1)
        grad_apn[:, 2] = (norm * ml_batch).sum(dim=1).sum(dim=1)

        #del mx_batch,my_batch,ml_batch,mx,my,ml,x,y,norm
        ##print 'grad apn: ', grad_apn
        print(" [*] Backward DIY calculated")
        return grad_input, grad_apn`
Thank you very much for your suggestions.

**Comment by jeong-tae** [comment]
Does backward operation should be computed in Variable?
hm... As far as i know, variable try to save all operation and gradient that you compute. In 0.4, i still try to distinct between variable and tensor.
What about change the Variable to Tensor in backward? did you tried?

**Comment by yuleichin** [comment]
@jeong-tae Yeah. I tried your code at the beginning. But it breaks down with an error telling that norm*mx_batch is not permitted because Variable cannot multiply with Tensor. Then, I print the type of grad_output in backward(). I find that grad_output is a Variable. Thus, we should use Variable in the backward process because the operation on Variable will be recorded into the graph.
